sendrecv/gst-rust/src/main.rs

```diff
@@ -0,0 +1,426 @@
+    offer: gstreamer_webrtc::WebRTCSessionDescription,
+    out: &ws::Sender,
+) {
+    if app_state.load(Ordering::Relaxed) < (AppState::PeerCallNegotiating as usize) {
+        // TODO signal and cleanup
+        panic!("Can't send offer, not in call");
+    };
+    let message = json!({
+          "sdp": {
+            "type": "offer",
+            "sdp": sdp_message_as_text(offer).unwrap(),
+          }
+        });
+    out.send(message.to_string()).unwrap();
+}
+
+fn on_offer_created(
+    app_state: &Arc<AtomicUsize>,
+    webrtc: gst::Element,
+    promise: &gst::Promise,
+    out: &ws::Sender,
+) {
+    assert_eq!(
+        app_state.load(Ordering::Relaxed),
+        AppState::PeerCallNegotiating as usize
+    );
+    let reply = promise.get_reply().unwrap();
+
+    let offer = reply
+        .get_value("offer")
+        .unwrap()
+        .get::<gstreamer_webrtc::WebRTCSessionDescription>()
+        .expect("Invalid argument");
+    let promise = gst::Promise::new();
+    webrtc
+        .emit(
+            "set-local-description",
+            &[&offer.to_value(), &promise.to_value()],
+        )
+        .unwrap();
+    send_sdp_offer(app_state, offer, out)
+}
+
+fn on_negotiation_needed(
+    app_state: &Arc<AtomicUsize>,
+    values: &[glib::Value],
+    out: &ws::Sender,
+) -> Option<glib::Value> {
+    app_state.store(AppState::PeerCallNegotiating as usize, Ordering::Relaxed);
+
+    let webrtc = values[0].get::<gst::Element>().expect("Invalid argument");
+    let webrtc_clone = webrtc.clone();
+    let out_clone = out.clone();
+    let app_state_clone = app_state.clone();
+    let promise = gst::Promise::new_with_change_func(move |promise: &gst::Promise| {
+        on_offer_created(&app_state_clone, webrtc, promise, &out_clone);
+    });
+    let options = gst::Structure::new_empty("options");
+    webrtc_clone
+        .emit("create-offer", &[&options.to_value(), &promise.to_value()])
+        .unwrap();
+    None
+}
+
+fn handle_media_stream(pad: &gst::Pad, pipe: &gst::Element, convert_name: &str, sink_name: &str) {
+    println!(
+        "Trying to handle stream with {} ! {}",
+        convert_name, sink_name,
+    );
+    let q = gst::ElementFactory::make("queue", None).unwrap();
+    let conv = gst::ElementFactory::make(convert_name, None).unwrap();
+    let sink = gst::ElementFactory::make(sink_name, None).unwrap();
+
+    if convert_name == "audioconvert" {
+        let resample = gst::ElementFactory::make("audioresample", None).unwrap();
+        pipe.clone()
+            .dynamic_cast::<gst::Bin>()
+            .unwrap()
+            .add_many(&[&q, &conv, &resample, &sink])
+            .unwrap();
+        q.sync_state_with_parent().unwrap();
+        conv.sync_state_with_parent().unwrap();
+        resample.sync_state_with_parent().unwrap();
+        sink.sync_state_with_parent().unwrap();
+        gst::Element::link_many(&[&q, &conv, &resample, &sink]).unwrap();
+    } else {
+        pipe.clone()
```

Comment:
You could move this `dynamic_cast` outside the `if` and have it only once then, storing the `bin` in a local variable